### ISSUE TITLE
Native image (Windows) 3x slower than JIT version when indexing objects in memory

### ISSUE THREAD

**Issue by Verdoso** [oracle/graal#5968]
**Describe the issue**
I created a Windows native version from a Spring Boot application that, upon startup, deserialised a couple of thousands objects into memory and then indexes  them using [CQEngine](https://github.com/npgall/cqengine). I noticed that the application was starting more to startup, so I managed to reduce the working case to the initial startup when deserialising the objects and indexing them.

**Steps to reproduce the issue**
I created a test case in this project: https://github.com/Verdoso/graalvm-test/. Basically, on an "x64 Native Tools Command Prompt":
1. **git clone --depth 1 https://github.com/Verdoso/graalvm-test**
2. **set JAVA_HOME=%GRAAL_HOME%**
3. **mvn -Pnative native:compile**
4. cd target
5. Java version: **%GRAAL_HOME%\bin\java -jar graalvm-test-0.0.1-SNAPSHOT.jar**
6. Native version: **graalvm_test.exe** 

**Describe GraalVM and your environment:**
- openjdk version "17.0.6" 2023-01-17
- OpenJDK Runtime Environment GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13)
- OpenJDK 64-Bit Server VM GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13, mixed mode, sharing)
 - Windows 10 19045.2486
 - Architecture: **Intelx64**

Output from the Java version
```
...\graalvm-test\target>%GRAAL_HOME%\bin\java -jar graalvm-test-0.0.1-SNAPSHOT.jar
16:59:38.837 INFO  - StartupInfoLogger - Starting Application v0.0.1-SNAPSHOT using Java 17.0.6 with PID 10808 (...\graalvm-test\target\graalvm-test-0.0.1-SNAPSHOT.jar started by scidlj in ...\graalvm-test\target)
16:59:38.847 INFO  - SpringApplication - No active profile set, falling back to 1 default profile: "default"
16:59:42.127 INFO  - StartupInfoLogger - Started Application in 4.144 seconds (process running for 4.818)
16:59:42.146 INFO  - DataLoaderService - Reading data...
16:59:42.147 WARN  - EplCSVProcessor - Reading file: ...\graalvm-test\target\..\test.ser
16:59:43.828 WARN  - EplCSVProcessor - Backup LeÝdo con 58387 libros.
16:59:43.829 INFO  - DataLoaderService - Preparando 58387 libros de la descarga con fecha 2023-02-11T04:00:26
16:59:56.339 INFO  - BibliotecaService - Libros a±adidos.
16:59:56.339 INFO  - BibliotecaService - Recopilando autores...
16:59:57.526 INFO  - BibliotecaService - Autores recopilados.
16:59:57.527 INFO  - BibliotecaService - Recopilando generos...
16:59:57.565 INFO  - BibliotecaService - GÚneros recopilados.
16:59:57.565 INFO  - BibliotecaService - Recopilando idiomas...
16:59:57.592 INFO  - BibliotecaService - Idiomas recopilados.
16:59:57.592 INFO  - BibliotecaService - Books: 58387
16:59:57.594 INFO  - DataLoaderService - Test finished in ~15s
```
Output from the native version
```
...\graalvm-test\target>graalvm_test.exe

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v3.0.2)

18:01:51.976 INFO  - StartupInfoLogger - Starting AOT-processed Application using Java 17.0.6 with PID 12732 (...\graalvm-test\target\graalvm_test.exe started by scidlj in ...\graalvm-test\target)
18:01:51.977 INFO  - SpringApplication - No active profile set, falling back to 1 default profile: "default"
18:01:52.264 WARN  - JvmGcMetrics - GC notifications will not be available because MemoryPoolMXBeans are not provided by the JVM
18:01:52.297 INFO  - StartupInfoLogger - Started Application in 0.362 seconds (process running for 0.37)
18:01:52.298 INFO  - DataLoaderService - Reading data...
18:01:52.298 WARN  - EplCSVProcessor - Reading file: ...\graalvm-test\target\..\test.ser
18:01:54.134 WARN  - EplCSVProcessor - Backup LeÝdo con 58387 libros.
18:01:54.134 INFO  - DataLoaderService - Preparando 58387 libros de la descarga con fecha 2023-02-11T04:00:26
18:02:33.660 INFO  - BibliotecaService - Libros a±adidos.
18:02:33.661 INFO  - BibliotecaService - Recopilando autores...
18:02:39.009 INFO  - BibliotecaService - Autores recopilados.
18:02:39.009 INFO  - BibliotecaService - Recopilando generos...
18:02:39.053 INFO  - BibliotecaService - GÚneros recopilados.
18:02:39.053 INFO  - BibliotecaService - Recopilando idiomas...
18:02:39.076 INFO  - BibliotecaService - Idiomas recopilados.
18:02:39.076 INFO  - BibliotecaService - Books: 58387
18:02:39.077 INFO  - DataLoaderService - Test finished in ~46s
```

All in all, in this test indexing of the 58K books takes 13s (Java) vs 39s (native), indexing authors takes 1.2s (Java) vs 5.4s (native) etc. I am wondering if is there something I'm doing wrong or if I could get some more info. The documentation I read said that debug information in the native was only available for *nix and this is a Windows binary so, any pointers would be appreciated.

All that those pieces of code do is add objects to a CQEngine ConcurrentIndexedCollection, that probably uses reflection to create its indexes around the attributes. Maybe so much reflection is the problem?

Not complaining, I wanted to share my experience it with a reproducible case, if you consider it is a problemand want to look at it.

In any case, I was able to generate the binary of my little utility and the users that are not able to upgrade to the latest version of Java can now use the native version while I use the latest Java, so they are happy and I'm happy. So, thank you!

**Comment by fniephaus** [comment]
Thanks, @Verdoso. The performance results are somewhat expected at the moment but we are constantly improving performance. For comparison, could you try building with GraalVM EE and see how that improves your benchmark? In a second step, use [PGO](https://www.graalvm.org/22.3/reference-manual/native-image/guides/optimize-native-executable-with-pgo/), which should further improve the throughput of your benchmark. If you could share some numbers, that would be great. If those numbers show something unexpected, we may indeed be looking at a bug (e.g., some missing intrinsic).

**Comment by Verdoso** [comment]
Changint GraalCE for GraalEE, same data, same machine
```
19:29:13.961 INFO  - DataLoaderService - Test finished in ~30s
```
Next step, GraalEE with PGO :).

**Comment by Verdoso** [comment]
Using a PGO optimized binary, startup remains at 27-28s...
Maybe it cannot help much in this specific case?

**Comment by fniephaus** [comment]
So you build with `--pgo-instrument`, then ran your benchmark, and then re-build it with `--pgo default.iprof` and then, the benchmark only runs slightly faster? Interesting result, but of course possible.

I didn't check what type of data you are processing in the benchmark, but if it can be pre-computed at build-time, you could freeze the result into the executable for much better startup. [Here](https://www.graalvm.org/22.3/reference-manual/native-image/basics/#build-time-vs-run-time) is more info if you're interested in this.

**Comment by Verdoso** [comment]
Yes, I built it using the two steps. When I run the instrumented executable it took 60s, normal, and generated the default.iprof. Then built again using that file.
Is there any way I can verify the executable is "pgo optimized"? Just to make sure.
In this case, the data cannot be pre-computed at build time, as the app is supposed to read new data from the network and index it. I just added a snapshot to test it using a controlled data set.

Anyway, it's good to know that some optimizations exist. I'll leave the project around to test against future versions and see how it evolves.
Thanks!

**Comment by fniephaus** [comment]
> Is there any way I can verify the executable is "pgo optimized"? Just to make sure.

I don't think there is, but that's something we probably should add to the build output. I'll track this internally.

---

We are working on more enhancements for PGO, so please feel free to update this with numbers for newer releases.

**Comment by Verdoso** [comment]
To add some more info, 
Tested on a different machine using Ubuntu under WSL 
Pure Java -> Test finished in ~12s
Native -> Test finished in ~31s

I'll try to add the EE and PGO optimized versions.
Cheers!
